Modules/selectmodule.c

```diff
@@ -554,7 +555,14 @@ poll_poll(pollObject *self, PyObject *args)
             return NULL;
         }
 
-        deadline = _PyTime_GetMonotonicClock() + timeout;
+        if (ms >= 0)
+            deadline = _PyTime_GetMonotonicClock() + timeout;
+        else    /* On many OSes timeout must be -1 or INFTIM, issue 31334 */
+#ifdef INFTIM
```

Comment:
Can't we do this in the following block?

```c
if (timeout_obj == NULL || timeout_obj == Py_None) {
    timeout = -1;
#ifdef INFTIM
    ms = INFTIM;
#else
    ms = -1;
#endif
    deadline = 0;
}
```

Then we can just change

```c
deadline = _PyTime_GetMonotonicClock() + timeout;
```

to

```c
if (ms >= 0) {
    deadline = _PyTime_GetMonotonicClock() + timeout;
}
```

Reply:
I don't see how what you're proposing can fix the issue. When a user passes a negative `timeout` parameter it enters the `else` block at line 542, so `ms` will be negative and the system call will still fail. The problem is not the `deadline` but the `ms` variable.

But your comment made me realise that I should add the `#ifdef` to the `if` block too, for coherence and in case an OS will ever have `INFTIM != -1`. To be DRY I'll move the `#ifdef` after the `else` block.